ext4/tar2ext4/tar2ext4.go

```diff
@@ -195,11 +222,404 @@ func Convert(r io.Reader, w io.ReadWriteSeeker, options ...Option) error {
+	}
+
+	// write the protectiveMBR at the beginning of the disk
+	_, err = w.Seek(0, io.SeekStart)
+	if err != nil {
+		return fmt.Errorf("failed to seek file: %v", err)
+	}
+	pMBR := gpt.ProtectiveMBR{
+		BootCode:               [440]byte{},
+		UniqueMBRDiskSignature: 0,
+		Unknown:                0,
+		PartitionRecord:        [4]gpt.PartitionMBR{},
+		Signature:              gpt.ProtectiveMBRSignature,
+	}
+
+	// See gpt package for more information on the max value of the size in LBA
+	sizeInLBA := uint32(alternateHeaderLBA)
+	if alternateHeaderLBA >= uint64(gpt.ProtectiveMBRSizeInLBAMaxValue) {
+		sizeInLBA = uint32(gpt.ProtectiveMBRSizeInLBAMaxValue)
+	}
+
+	pMBR.PartitionRecord[0] = gpt.PartitionMBR{
+		BootIndicator: 0,
+		StartingCHS:   gpt.ProtectiveMBRStartingCHS,
+		OSType:        gpt.ProtectiveMBRTypeOS,
+		EndingCHS:     calculateEndingCHS(sizeInLBA),
+		StartingLBA:   gpt.PrimaryHeaderLBA, // LBA of the GPT header
+		SizeInLBA:     sizeInLBA,            // size of disk minus one is the alternate header LBA
+	}
+
+	// write the protectiveMBR
+	if err := binary.Write(w, binary.LittleEndian, pMBR); err != nil {
+		return fmt.Errorf("failed to write backup header with: %v", err)
+	}
+
+	// write partition entries
+	_, err = w.Seek(int64(gpt.PrimaryEntryArrayLBA*compactext4.BlockSizeLogical), io.SeekStart)
+	if err != nil {
+		return fmt.Errorf("failed to seek file: %v", err)
+	}
+
+	for _, e := range entries {
+		if err := binary.Write(w, binary.LittleEndian, e); err != nil {
+			return fmt.Errorf("failed to write backup entry array with: %v", err)
+		}
+	}
+
+	// only calculate the checksum for the actual partition array, do not include reserved bytes if any
+	entriesCheckSum, err := getChecksumPartitionEntryArray(w, gpt.PrimaryEntryArrayLBA, uint32(sizeOfEntryArrayBytes))
+	if err != nil {
+		return err
+	}
+
+	// write primary gpt header
+	_, err = w.Seek(int64(gpt.PrimaryHeaderLBA*compactext4.BlockSizeLogical), io.SeekStart)
+	if err != nil {
+		return fmt.Errorf("failed to seek file: %v", err)
+	}
+
+	hGPT := gpt.Header{
+		Signature:                gpt.HeaderSignature,
+		Revision:                 gpt.HeaderRevision,
+		HeaderSize:               gpt.HeaderSize,
+		HeaderCRC32:              0, // set to 0 then calculate crc32 checksum and replace
+		ReservedMiddle:           0,
+		MyLBA:                    uint64(gpt.PrimaryHeaderLBA), // LBA of this header
+		AlternateLBA:             alternateHeaderLBA,
+		FirstUsableLBA:           firstUseableLBA,
+		LastUsableLBA:            lastUseableLBA,
+		DiskGUID:                 diskGUID,
+		PartitionEntryLBA:        uint64(gpt.PrimaryEntryArrayLBA), // right after this header
+		NumberOfPartitionEntries: uint32(len(multipleReaders)),
+		SizeOfPartitionEntry:     gpt.HeaderSizeOfPartitionEntry,
+		PartitionEntryArrayCRC32: entriesCheckSum,
+		ReservedEnd:              [420]byte{},
+	}
+	hGPT.HeaderCRC32, err = getHeaderChecksum(hGPT)
```

Comment:
This is currently writing the block into memory, computing a checksum, updating the original block, then writing it again to the final destination. I'd suggest instead we wrap the `Writer` in a type that computes a running checksum as it writes out bytes to the final destination, then we can seek back to the checksum offset and write it after the fact.

Reply:
We need to use `binary.Write` (afaik) to ensure the correct byte order. `binary.Write` will essentially read the struct into a buffer in memory and then write that to disk. If we made a wrapper around `binary.Write`, we'd still need to read the data into a buffer to create the checksum since we can't access the one created in `binary.Write` and then pass that back through the `binary.Write`, which will then read it into memory again before writing it to the disk. So, with this route we'd still be essentially reading the data into memory twice. Unless we want to handle writing to disk at a lower level, requiring us to manually deal with the byte order, but I don't think that's worth it.

Reply:
When dealing with types where the C ABI matches what we need (e.g. is little-endian, doesn't have any undesirable padding), we can actually get away with just writing the Go type out directly (by getting a pointer to it and its size) instead of using `binary.Write`.

If we didn't have the reserved field, we could create a `Writer` type that records the checksum for us, and pass that to `binary.Write`, and it would work. We would still need to either write the whole header again, or seek to the header offset in the file and write just the checksum value out. Does that make sense?

All that said, I don't really have a problem with how you're doing it, as the data is only 512 bytes. I just wanted you to consider alternate approaches, and think about not buffering in memory when we can avoid it.

Reply:
I still don't understand where reading the data into memory twice would be needed, so happy to discuss further.